simple_network_sim/sns_logger.py

```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+"""Provides logging support for the simple_network_sim module.
+
+This approach define a package-level logger that works as follows:
+
+1. Import sns_logger from any package module - this will instantiate
+a new logging.Logger instance as sns_logger.logger and use the
+setup_logger() function to set the default formatting. From the point
+of first import, the sns_logger.logger instance will be available
+for logging.
+
+2. To modify logger configuration (e.g. on the basis of CLI options),
+call setup_logger() with an argparse.Namespace that includes the
+fields 'logfile', 'verbose', and 'debug'. This will change global
+settings for this logger. For example, setting a logfile in the entry
+point for a script will make that logfile available to all logger
+calls (where the sns_logger.logger is used). Calling setup_logger()
+should - for consistency - be done before any information is written
+to the logger.
+
+To use the package-level logger, you might use
+
+from . import sns_logger
+logger = sns_logger.logger
+
+or
+
+from simple_network_sim import sns_logger
+logger = sns_logger.logger
+
+NOTE: By default, the logger writes to sys.stderr and only reports
+logger.WARNING and above.
+"""
+
+import logging
+import logging.config
+import sys
+
+from argparse import Namespace
+from pathlib import Path
+from typing import Optional
+
+def setup_logger(args: Optional[Namespace] = None) -> None:
+    """Configure the package-wide logging.Logger instance.
+    
+    :param args: argparse.Namespace
+        args.logfile (pathlib.Path) is used to create a logfile if present
+        args.verbose and args.debug control logging level to sys.stderr
+    """
+    # Dictionary to define logging configuration
+    logconf = {
+        "version": 1,
+        "disable_existing_loggers": False,
+        "formatters": {
+            "standard": {
+                "format": "[%(asctime)s] [%(name)s] [%(levelname)s] %(message)s",
+                },
+            },
+        "handlers": {
+            "stderr": {
+                "level": "WARNING",
```

Comment:
I feel like INFO should be the default. For instance, the parameters are printed in the INFO level, and you probably want to see that after each run. It's common for schedulers to run a command and then give you access to the stdout/stderr. But even when you just run locally, it'll probably be useful to be able to go back in history and what all the parameters were.

Reply:
I don't think this gets in the way of what you're asking for.

`DEBUG` is the lowest reportable state for the logger, but the handlers defined in `"handlers"` override this. Slightly above that in the code (line 61) the lowest reported status for the default loghandler that writes to STDERR* is defined: this is `WARNING` (by default). That `WARNING` status is overridden if `verbose` or `debug` options are selected in the CLI (lines 96-101).

The overall behaviour is that:

- by default, only `WARNING` and above are written to STDERR
- if `verbose` is flagged, `INFO` and above are written to STDERR
- if `debug` is flagged, `DEBUG` and above are written to STDERR
- if `logfile` is provided, `INFO` and above are written to a file whose name is provided (this is `DEBUG` and above if `debug` is flagged).

This provides some flexibility in providing user messages through the logger:

- fatal problems can be flagged with `logger.error()` (always visible)
- worrying problems that should always be seen can be flagged with `logger.warning()` (always visible)
- useful but nonessential messages can be controlled with `verbose` (default: off)
- messages only useful for monitoring the simulation/run can be controlled with `debug` (default: off).

I'm not wedded to default settings, but I hoped you'd find this flexibility useful.

*The default loghandler writes to STDERR because this is where I'd expect error/information messages to be - STDOUT tends to be reserved for output you might pipe elsewhere.

Reply:
I understand. My suggestion was that INFO was written by default (without the need of passing verbose). I had initially commented in the wrong line, that may have caused some confusion.